File: PHP/NScAPIWrapper.php
<?php
/**
 * This class implements the different endpoints as documented
 * @author Franclin <franclin.foping@nitrosell.net>
 * 
 */
class NScAPIWrapper {

  private $sKey = null; ///< This variable will hold the private key of the user
  private $sUserID = null; ///< This variable will hold the user ID
  private $bDebug = false; ///< This boolean is useful for debugging purpose
  private $responseInfo = array(); ///< This array will be used to store the response from curl   
  private $user_agent = null; ///< This variable will contain the user agent description 
  private $webstoreUrl = null; ///< This variable will contain the URL of the WebStore

  /**
   * This is the constructor of our class. Its job is to set the member variable to their correct values
   * @param type $key  This variable will hold the private key of the user
   * @param type $userid  This variable will hold the user ID
   * @param type $webstoreUrl  This variable will contain the URL of the WebStore
   */
  public function __construct($key, $userid, $webstoreUrl) {
    $this->sKey = $key;
    $this->sUserID = $userid;
    $this->webstoreUrl = $webstoreUrl;

    $this->user_agent = "Testing user agent";
  }

  /**
   * This function sets the private key
   * @param type $key  This is the new private key
   */
  public function setPrivateKey($key) {
    $this->sKey = $key;
  }

  /**
   * This function returns the current user ID 
   * @return the user id 
   */
  public function getUserID() {
    return $this->sUserID;
  }

  /**
   * This helper method is used to process the actual HTTP request.
   * We used Curl here
   * @param type $sUrl this is the URL where the request will be sent
   * @param type $postargs this is the query string to be sent as part of the HTTP request
   * @return the body of the response from the HTTP request
   */
  private function processRequest($sUrl, $postargs = false) {
    $ch = curl_init($sUrl);

    if ($postargs) {
      $arrPostData = array();

      foreach($postargs as $key => $value) {
        $arrPostData[] = $key . '=' . $value;
      }     

      $arrPostData = implode("&", $arrPostData);

      // setting options for curl
      curl_setopt($ch, CURLOPT_POST, true);
      curl_setopt($ch, CURLOPT_POSTFIELDS, $arrPostData);
      curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-type: application/x-www-form-urlencoded'));//, 'Content-length: 100'));
    }

    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_VERBOSE, false);
    curl_setopt($ch, CURLOPT_NOBODY, false);

    if ($this->bDebug) {
      curl_setopt($ch, CURLOPT_HEADER, true);
    } else {
      curl_setopt($ch, CURLOPT_HEADER, false);
    }

    curl_setopt($ch, CURLOPT_USERAGENT, $this->user_agent);

    $response = curl_exec($ch);

    $this->responseInfo = curl_getinfo($ch);

    curl_close($ch);

    if( $this->bDebug) {
      $debug = preg_split("#\n\s*\n|\r\n\s*\r\n#m", $response);
      echo'<pre>' . $debug[0] . '</pre>';
      exit;
    };
    return $response;
 }

 /**
  * This endpoint returns the customer's data
  * @param type $email this is the email of the customer whose details are being requested
  * @return string the JSON encoded string containing the customer data
  */
 public function getCustomerDetails($email) {
   $requestUrl = "https://api.internal.nitrosell.com/".$this->webstoreUrl."/v1/customer.json";

   $time = time();

   $arrData = array(
       'email' => $email,
       'userid' => $this->sUserID,
       'time' => $time
   );

   $hash = hash_hmac("md5", join(".", $arrData), $this->sKey);

   $queryString = "?email=".$email."&hash=".$hash."&userid=".$this->sUserID."&time=".$time;

   return $this->processRequest($requestUrl . $queryString);
 }

 /**
  * This function returns the set of shipping options for the basket
  * @param type $postargs an associative array containing the basket data
  * @return string a JSON encoded array of the shipping options
  */
 public function getShippingOptions($postargs) {
   $requestUrl = "https://api.internal.nitrosell.com/".$this->webstoreUrl."/v1/shipping.json";
   
   $arrRawPostParameters = array();
   $arrParameters = array();
   parse_str($postargs, $arrRawPostParameters);

   foreach ($arrRawPostParameters as $key => $value) {
     $arrParameters[$key] = urldecode($value);
   }

   $arrParameters['hash'] = hash_hmac("md5", join(".", $arrParameters), $this->sKey);
   
   return $this->processRequest($requestUrl, $arrParameters);
 }
 
 
 /**
  * This endpoint returns the amount of applicable tax for the basket
  * @param type $postargs the basket data
  * @return string the amount of tax in a JSON encoded string
  */
 public function getTaxForAnOrder($postargs) {
   $requestUrl = "https://api.internal.nitrosell.com/".$this->webstoreUrl."/v1/tax.json";
   $arrRawPostParameters = array();
   $arrParameters = array();
   parse_str($postargs, $arrRawPostParameters);

   foreach ($arrRawPostParameters as $key => $value) {
     $arrParameters[$key] = urldecode($value);
   }
   
   $arrParameters['hash'] = hash_hmac("md5", join(".", $arrParameters), $this->sKey);
   
   return $this->processRequest($requestUrl, $arrParameters);
 }


 /**
  * This function performs the actual order insertion into the database
  * @param array $param an associative array containing the basket details
  * @return string 
  */
 public function insertWebOrders($param) {
   $requestUrl = "https://api.internal.nitrosell.com/".$this->webstoreUrl."/v1/order.json";
   $arrRawPostParameters = array();
   $arrParameters = array();
   parse_str($param, $arrRawPostParameters);

   foreach ($arrRawPostParameters as $key => $value) {
     $arrParameters[$key] = urldecode($value);
   }
   
   $arrParameters['hash'] = hash_hmac("md5", join(".", $arrParameters), $this->sKey);
   
   return $this->processRequest($requestUrl, $arrParameters);
 }
 
 /**
  * This endpoint computes and returns the amount of tax and shipping options applicable to a basket
  * @param type $postargs this is an associate array containing the description of the basket
  * @return string a JSON encoded string containing both the amount of tax and shipping options applicable to a basket
  */
 public function getShippingAndTax($postargs) {
   $requestUrl = "https://api.internal.nitrosell.com/".$this->webstoreUrl."/v1/shippingandtax.json";
   $arrRawPostParameters = array();
   $arrParameters = array();
   parse_str($postargs, $arrRawPostParameters);

   foreach ($arrRawPostParameters as $key => $value) {
     $arrParameters[$key] = urldecode($value);
   }
   
   $arrParameters['hash'] = hash_hmac("md5", join(".", $arrParameters), $this->sKey);
   
   return $this->processRequest($requestUrl, $arrParameters);
 }

}
?>

File: Java/NScAPIWrapper.java
/*
 * This file contains the implementation of the main class for the API
 * The class contains five public methods to access each endpoint of the API and a handful of private methods
 * Each method is documented below
 * @author franclin
 * @date 04/02/2013
 */
package apisamplecode;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 *
 * @author franclin
 * @date 04/02/2013
 */
public class NScAPIWrapper {

  private String _key;
  private String _userid;
  private String _webstoreUrl;
  private URL _url = null;
  private HttpURLConnection _connection = null;
  private SecretKeySpec _keySpec = null;
  private Mac _mac = null;

  /**
   * 
   * This is the constructor of the class
   * Three parameters are needed to build an object of this type
   * 
   * @param key this represents your private. You should contact the NitroSell Support Team to request one. 
   * @param userid this represents your User ID as assigned by the retailer to which you would like to query his data
   * @param webstoreUrl this is the URL of the WebStore that you would like to query
   */
  public NScAPIWrapper(String key, String userid, String webstoreUrl) {
    this._key = key;
    this._userid = userid;
    this._webstoreUrl = webstoreUrl;

    _keySpec = new SecretKeySpec(this._key.getBytes(), "HmacMD5");
    try {
      _mac = Mac.getInstance("HmacMD5");
      _mac.init(_keySpec);
    } catch (Exception e) {
      System.out.println(e.getMessage());
    }
    //this._connection = new HttpURLConnection();
  }
  
  /**
   * This helper method builds an HTTP POST request given the parameters and the action to be taken
   * 
   * - The parameters is a hash table containing all the details required to make the request
   * - The action is the resource being requested. It could be either shipping, shippingandtax, tax or order
   * @param parameters
   * @param action 
   */
  private void buildRequestAndExecute(HashMap parameters, String action) {
    String toencode = "";
    Set set = parameters.entrySet();
    Iterator iter = set.iterator();

    while (iter.hasNext()) {
      Map.Entry e = (Map.Entry) iter.next();
      toencode += e.getValue() + ".";
    }

    toencode = toencode.substring(0, toencode.length() - 1);
    byte[] rawHmac = _mac.doFinal(toencode.getBytes());

    BigInteger hash = new BigInteger(1, rawHmac);
    String hmac = hash.toString(16);

    if (hmac.length() % 2 != 0) {
      hmac = "0" + hmac;
    }
    
    parameters.put("hash", hmac);

    try {
      this._url = new URL("http://api.nitrosell.com/" + this._webstoreUrl + "/v1/"+ action +".json");
      _connection = (HttpURLConnection) _url.openConnection();
      _connection.setRequestMethod("POST");
      _connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
      _connection.setRequestProperty("Content-Length", Integer.toString(this.urlEncodeUTF8(parameters).length()));
      _connection.setDoOutput(true);
      _connection.setUseCaches(false);
      _connection.setDoInput(true);
      _connection.setAllowUserInteraction(false);

      //Send request
      DataOutputStream wr = new DataOutputStream(_connection.getOutputStream());
      wr.writeBytes(this.urlEncodeUTF8(parameters));
      wr.flush();
      wr.close();

      // Get response
      InputStream is = _connection.getInputStream();
      BufferedReader br = new BufferedReader(new InputStreamReader(is));
      String line;

      StringBuilder response = new StringBuilder();

      while ((line = br.readLine()) != null) {
        response.append(line);
        response.append("\r");
      }

      br.close();

      System.out.println("Response " + response.toString());
    } catch (Exception e) {
      System.out.println(e.getMessage());
    }
  }
  
  
  /**
   * This helper function returns the URL Encoded UTF-8 version of the input string
   * @param s the string to be encoded
   * @return 
   */
  private String urlEncodeUTF8(String s) {
    try {
      return URLEncoder.encode(s, "UTF-8");
    } catch (Exception e) {
      throw new UnsupportedOperationException(e);
    }
  }
  
  /**
   * This helper method returns the query string used that will be passed in the body of the POST request   * 
   * @param map
   * @return string 
   */
  private String urlEncodeUTF8(Map<?, ?> map) {
    StringBuilder sb = new StringBuilder();
    
    for (Map.Entry<?, ?> entry : map.entrySet()) {
      if (sb.length() > 0) {
        sb.append("&");
      }
      
      sb.append(String.format("%s=%s", urlEncodeUTF8(entry.getKey().toString()), urlEncodeUTF8(entry.getValue().toString())));      
    }
    
    return sb.toString();
  }
  
  
  /**
   * This method returns the user ID 
   * 
   * @return 
   */
  public String getUserID() {
    return this._userid;
  }

  /**
   * This method is the endpoint to get the customer details
   * @param email 
   */
  public void getCustomerDetails(String email) {
    Date time = new Date();
    long timestamp = time.getTime() / 1000;

    try {

      String message = email + "." + this._userid + "." + timestamp;

      byte[] rawHmac = _mac.doFinal(message.getBytes());

      BigInteger hash = new BigInteger(1, rawHmac);
      String hmac = hash.toString(16);

      if (hmac.length() % 2 != 0) {
        hmac = "0" + hmac;
      }

      this._url = new URL("http://api.nitrosell.com/" + this._webstoreUrl + "/v1/customer.json" + "?email=" + email + "&hash=" + hmac + "&userid=" + this._userid + "&time=" + timestamp);
      this._connection = (HttpURLConnection) _url.openConnection();

      InputStream is = this._connection.getInputStream();
      BufferedReader br = new BufferedReader(new InputStreamReader(is));

      String line;
      StringBuilder response = new StringBuilder();

      while ((line = br.readLine()) != null) {
        response.append(line);
        response.append("\r");
      }

      br.close();
      System.out.print(response.toString());

    } catch (Exception e) {
      System.out.println("Error : " + e.getMessage());
      _connection.disconnect();
    } finally {
      if (_connection != null) {
        _connection.disconnect();
      }
    }
  }
  
  /**
   * This method is the endpoint to get the list of shipping options
   * The parameters of the request are passed in parameters tables
   * @param parameters 
   */

  public void getShippingOptions(HashMap parameters) {
    this.buildRequestAndExecute(parameters, "shipping");   
  }
  
  /**
   * This method is the endpoint to get the list of shipping options and any applicable tax 
   * The parameters of the request are passed in parameters tables
   * @param parameters 
   */
  
  public void getShippingAndTax(HashMap parameters) {
    
    this.buildRequestAndExecute(parameters, "shippingandtax");   
  }
  
  /**
   * This method is the endpoint to get the tax for a given order
   * The parameters of the request are passed in parameters tables
   * @param parameters 
   */
  
  public void getTaxForAnOrder(HashMap parameters) {
    
    this.buildRequestAndExecute(parameters, "tax");
   
  }
  
  
  /**
   * This method is the endpoint to insert an order into the WebStore
   * The parameters of the request are passed in parameters tables
   * @param parameters 
   */
  
  public void insertWebOrders(HashMap parameters) {
    this.buildRequestAndExecute(parameters, "order");
  }
}
